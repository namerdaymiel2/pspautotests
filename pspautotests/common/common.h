#undef main
#define main main2

extern int RUNNING_ON_EMULATOR;
