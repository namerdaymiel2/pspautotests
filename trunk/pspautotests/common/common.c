#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <pspdebug.h>
#include <pspthreadman.h>
#include <psploadexec.h>
#include <pspctrl.h>
#include <pspiofilemgr.h>
//#include <pspkdebug.h>

#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <sys/lock.h>
#include <sys/fcntl.h>
//#include "local.h"

PSP_MODULE_INFO("TESTMODULE", PSP_MODULE_USER, 1, 0);
PSP_MAIN_THREAD_ATTR(THREAD_ATTR_USER | THREAD_ATTR_VFPU);

#define EMULATOR_DEVCTL__GET_HAS_DISPLAY 0x00000001

unsigned int RUNNING_ON_EMULATOR = 0;
unsigned int HAS_DISPLAY = 1;

extern int test_main(int argc, char *argv[]);

FILE stdout_back = {NULL};
SceUID KprintfFd = 0;

static int writeStdoutHook(struct _reent *ptr, void *cookie, const char *buf, int buf_len) {
	char temp[1024 + 1];

	if (KprintfFd != 0) sceIoWrite(KprintfFd, buf, buf_len);

	if (buf_len < sizeof(temp)) {
		if (HAS_DISPLAY) {
			memcpy(temp, buf, buf_len);
			temp[buf_len] = 0;

			//Kprintf("%s", temp);
			pspDebugScreenPrintf("%s", temp);
		}
	}
	
	if (stdout_back._write != NULL) {
		return stdout_back._write(ptr, cookie, buf, buf_len);
	} else {
		return buf_len;
	}
}

void test_begin() {
	if (HAS_DISPLAY) {
		pspDebugScreenInit();
	}

	if (RUNNING_ON_EMULATOR) {
		fclose(stdout);
		stdout = fmemopen(alloca(4), 4, "wb");
		stdout_back._write = NULL;
		
		//stderr = stdout;
		setbuf(stdout, NULL);
	} else {
		freopen("ms0:/__testoutput.txt", "wb", stdout);
		freopen("ms0:/__testerror.txt", "wb", stderr);
		stdout_back._write = stdout->_write;
	}
	stdout->_write = writeStdoutHook;
	
	setbuf(stderr, NULL);
}

void test_end() {
	fflush(stdout);
	fflush(stderr);

	if (!RUNNING_ON_EMULATOR) {
		SceCtrlData key;
		while (1) {
			sceCtrlReadBufferPositive(&key, 1);
			if (key.Buttons & PSP_CTRL_CROSS) break;
		}
	}
	
	//fclose(stdout);
	sceKernelExitGame();
	
	exit(0);
}

int test_psp_exit_callback(int arg1, int arg2, void *common) {
	exit(0);
	return 0;
}

int test_psp_callback_thread(SceSize args, void *argp) {
	int cbid;
	cbid = sceKernelCreateCallback("Exit Callback", test_psp_exit_callback, NULL);
	sceKernelRegisterExitCallback(cbid);
	sceKernelSleepThreadCB();
	return 0;
}

int test_psp_setup_callbacks(void) {
	int thid = 0;
	thid = sceKernelCreateThread("update_thread",  test_psp_callback_thread, 0x11, 0xFA0, 0, 0);
	if (thid >= 0) sceKernelStartThread(thid, 0, 0);
	return thid;
}

//#define START_WITH "ms0:/PSP/GAME/virtual"

/*
void emitInt(int v) {
	asm("syscall 0x1010");
}

void emitFloat(float v) {
	asm("syscall 0x1011");
}

void emitString(char *v) {
	asm("syscall 0x1012");
}

void emitComment(char *v) {
	asm("syscall 0x1012");
}

void emitMemoryBlock(void *address, unsigned int size) {
	asm("syscall 0x1013");
}

void emitHex(void *address, unsigned int size) {
	asm("syscall 0x1014");
}
*/

int main(int argc, char *argv[]) {
	int retval = 0;

	KprintfFd = sceIoOpen("emulator:/Kprintf", O_WRONLY, 0777);
	
	RUNNING_ON_EMULATOR = (KprintfFd > 0);

	if (RUNNING_ON_EMULATOR) {
		sceIoDevctl("emulator:", EMULATOR_DEVCTL__GET_HAS_DISPLAY, NULL, 0, &HAS_DISPLAY, sizeof(HAS_DISPLAY));
	}

	//if (strncmp(argv[0], START_WITH, strlen(START_WITH)) == 0) RUNNING_ON_EMULATOR = 1;

	if (!RUNNING_ON_EMULATOR) {
		test_psp_setup_callbacks();
	}
	atexit(sceKernelExitGame);

	test_begin();
	{
		pspDebugScreenPrintf("RUNNING_ON_EMULATOR: %s - %s\n", RUNNING_ON_EMULATOR ? "yes" : "no", argv[0]);
		retval = test_main(argc, argv);
	}
	test_end();
	
	return retval;
}
