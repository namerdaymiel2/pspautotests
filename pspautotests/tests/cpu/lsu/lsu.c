#include <common.h>

#include <pspkernel.h>

///////////////////////////////////////////////////////////////////////////////
// LWL/LWR                                                                   //
///////////////////////////////////////////////////////////////////////////////

#define LWRL_DEF(TYPE, POS) \
	unsigned int TYPE##_##POS(unsigned int address) { \
		int ret; \
		asm volatile ( \
			"li $t0, 0x12345678\n" \
			#TYPE " $t0, " #POS "(%1)\n" \
			"move %0, $t0\n" \
			\
			: "=r"(ret) : "r" (address) \
		); \
		return ret; \
	}

#define LWR_DEF(POS) LWRL_DEF(lwr, POS)
#define LWL_DEF(POS) LWRL_DEF(lwl, POS)

#define LWRL_CALL(TYPE, POS) \
	printf(#TYPE "_" #POS " 0x%08X\n", TYPE##_##POS((unsigned int)&data[0]));

#define LWR_CALL(POS) LWRL_CALL(lwr, POS)
#define LWL_CALL(POS) LWRL_CALL(lwl, POS)

///////////////////////////////////////////////////////////////////////////////
// SWL/SWR                                                                   //
///////////////////////////////////////////////////////////////////////////////

#define SWRL_DEF(TYPE, POS) \
	unsigned int TYPE##_##POS(unsigned int address) { \
		int ret; \
		asm volatile ( \
			"li $t0, 0x12345678\n" \
			#TYPE " $t0, " #POS "(%1)\n" \
			"move %0, $t0\n" \
			\
			: "=r"(ret) : "r" (address) \
		); \
		return ret; \
	}

#define SWR_DEF(POS) SWRL_DEF(swr, POS)
#define SWL_DEF(POS) SWRL_DEF(swl, POS)

#define SWRL_CALL(TYPE, POS) \
	memcpy(data_copy, data, sizeof(data)); \
	TYPE##_##POS((unsigned int)&data_copy[0]); \
	printf(#TYPE "_" #POS " 0x%08X\n", *(unsigned int *)&data_copy[0]);

#define SWR_CALL(POS) SWRL_CALL(swr, POS)
#define SWL_CALL(POS) SWRL_CALL(swl, POS)


unsigned char data[16] = { 0xF0, 0xE1, 0xD2, 0xC3, 0xB4, 0xA5, 0x96, 0x87, 0x78, 0x69, 0x5A, 0x4B, 0x3C, 0x2D, 0x1E, 0x0F };
unsigned char data_copy[16];

LWR_DEF(0)
LWR_DEF(1)
LWR_DEF(2)
LWR_DEF(3)

LWL_DEF(0)
LWL_DEF(1)
LWL_DEF(2)
LWL_DEF(3)

SWR_DEF(0)
SWR_DEF(1)
SWR_DEF(2)
SWR_DEF(3)

SWL_DEF(0)
SWL_DEF(1)
SWL_DEF(2)
SWL_DEF(3)

void testAlignedLwlrSwlr() {
	LWR_CALL(0);
	LWR_CALL(1);
	LWR_CALL(2);
	LWR_CALL(3);
	printf("\n");

	LWL_CALL(0);
	LWL_CALL(1);
	LWL_CALL(2);
	LWL_CALL(3);
	printf("\n");

	SWR_CALL(0);
	SWR_CALL(1);
	SWR_CALL(2);
	SWR_CALL(3);
	printf("\n");

	SWL_CALL(0);
	SWL_CALL(1);
	SWL_CALL(2);
	SWL_CALL(3);
	printf("\n");
}

void testUnalignedLwlrSwlr() {
	printf("@TODO: testUnalignedLwlrSwlr\n");
}

int main(int argc, char *argv[]) {
	testAlignedLwlrSwlr();
	testUnalignedLwlrSwlr();

	return 0;
}