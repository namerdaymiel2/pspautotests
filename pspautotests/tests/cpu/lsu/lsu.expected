lwr_0 0xC3D2E1F0
lwr_1 0x12C3D2E1
lwr_2 0x1234C3D2
lwr_3 0x123456C3

lwl_0 0xF0345678
lwl_1 0xE1F05678
lwl_2 0xD2E1F078
lwl_3 0xC3D2E1F0

swr_0 0x12345678
swr_1 0x345678F0
swr_2 0x5678E1F0
swr_3 0x78D2E1F0

swl_0 0xC3D2E112
swl_1 0xC3D21234
swl_2 0xC3123456
swl_3 0x12345678